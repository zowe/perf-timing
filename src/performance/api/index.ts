export * from "./errors";
export * from "./interfaces";
export * from "./PerformanceApi";
