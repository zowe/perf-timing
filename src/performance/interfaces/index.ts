export * from "./IEnabled";
export * from "./IPerformanceTools";
