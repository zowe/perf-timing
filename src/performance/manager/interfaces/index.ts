export * from "./IEnabled";
export * from "./IPerformanceApi";
