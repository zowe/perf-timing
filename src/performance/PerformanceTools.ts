/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

// Imported for typings. This will not be reflected in the generated code.
import * as perfHooks from "perf_hooks";

import {IPerfEnabled} from "./interfaces";

// @TODO Separate file for the interfaces

interface IFunctionTimer {
    observer: perfHooks.PerformanceObserver;
    originalFunction: (...args: any[]) => any;
    totalDuration: number;
    totalCalls: number;
}

interface IMeaureTimer {
    observer: perfHooks.PerformanceObserver;
    measurements: IMeasurment[];
}

interface IMeasurment {
    name: string;
    startTime: number;
    startMarkName: string;
    endMarkName: string;
    duration: number;
}

export class PerformanceTools {

    // @TODO DOCUMENT
    private _functionTimers: Map<string,IFunctionTimer> = new Map();
    private _measureTimers: Map<string, number> = new Map();

    /**
     * This variable holds the import from the Node JS performance hooks
     * library.
     */
    private readonly _perfHooks: typeof perfHooks;

    constructor(private readonly _manager: IPerfEnabled) {
        // Check if performance utilities should be enabled.
        if(this._manager.isPerfEnabled) {
            // Delay the require so we don't waste resources when performance
            // isn't needed.
            this._perfHooks = require("perf_hooks");

            // process.on("exit", () => {
            //    console.log("NODE EXIT OCCURING...PRINTING METRICS");
            //    console.log("-------------------------------------");
            //    this.outputMetrics();
            // });
        }
    }

    public clearMarks(name?: string) {
        if (this._manager.isPerfEnabled) {
            this._perfHooks.performance.clearMarks(name);
        }
    }

    public mark(name: string) {
        if (this._manager.isPerfEnabled) {
            this._perfHooks.performance.mark(name);
        }
    }

    public measure(name: string, startMark: string, endMark: string) {
        if (this._manager.isPerfEnabled) {
            const obs = new this._perfHooks.PerformanceObserver((list, observer) => {
                console.log(list.getEntriesByName(name));

                observer.disconnect();
            });
            obs.observe({entryTypes: ["measure"]});

            this._perfHooks.performance.measure(name, startMark, endMark);
        }
    }

    public timerify(fn: (...args: any[]) => any, name?: string) {
        if (this._manager.isPerfEnabled) {
            // Check if we should use the function name or the passed name for tracking.
            if (name == null) {
                name = fn.name;
            }

            // Throw an error if the timer already exists in the map.
            if (this._functionTimers.has(name)) {
                throw new (require("./errors").TimerNameConflictError)(name);
            }

            // Create the observer and store it in the map.
            const observerObject: IFunctionTimer = {
                observer: undefined,
                originalFunction: fn,
                totalDuration: 0,
                totalCalls: 0
            };

            this._functionTimers.set(name, observerObject);

            // Create a function observer
            observerObject.observer = new this._perfHooks.PerformanceObserver((list) => {
                // const entries = list.getEntriesByName(fn.name);
                const entries = list.getEntriesByName(fn.name);

                for (const entry of entries) {
                    observerObject.totalDuration += entry.duration;
                    observerObject.totalCalls++;
                }
            });

            observerObject.observer.observe({entryTypes: ["function"], buffered: true});

            // Wrap the function in a timer
            return this._perfHooks.performance.timerify(fn);
        } else {
            // If performance is not enabled, we need to return the function
            // so original code doesn't get broken.
            return fn;
        }
    }

    /**
     * This method will close an observer that was opened through the {@link PerformanceTools#timerify}
     * function.
     *
     * @todo document
     */
    public untimerify(fn: ((...args: any[]) => any), name?: string) {
        if (this._manager.isPerfEnabled) {
            let timer = fn.name;

            // Extract the name of the function if necessary
            if (name != null) {
                timer = name;
            } else {
                // When timerifing(?) functions, node will prepend the function
                // name with timerified . This regex will strip out that name
                timer = /timerified (.*)/g.exec(timer)[1];
            }

            const timerRef = this._functionTimers.get(timer);

            if (timerRef !== undefined) {
                timerRef.observer.disconnect();
                return timerRef.originalFunction;
            } else {
                throw new (require("./errors").TimerDoesNotExistError)(timer);
            }
        } else {
            return fn;
        }
    }

    /**
     * Output raw performance metrics to a file. Should be the last call in execution.
     */
    public getMetrics(): string {
        if (this._manager.isPerfEnabled) {
            // @TODO All metrics should be stopped before reporting

            let output = "NODE EXIT OCCURING...PRINTING METRICS\n";
            output += "-------------------------------------\n\n";

            const timing = this._perfHooks.performance.nodeTiming;

            output += "Node Run Statistics\n";
            output += "-------------------\n";
            output += `Node Initialized in ${timing.nodeStart}ms\n`;
            output += `V8 Platform Initialized in ${timing.v8Start}ms\n`;
            output += `Process Bootstrapped in ${timing.bootstrapComplete}ms\n`;
            output += `Process Loop Started in ${timing.loopStart}ms\n`;
            output += `Process Loop Ended after ${timing.loopExit - timing.loopStart}ms\n`;
            output += `Total Process Duration: ${timing.duration}ms\n`;

            // All metrics should
            output += "\nFunction Timers\n";
            output += "---------------\n";

            const functionTimers = this._functionTimers.entries();

            for (const [key, value] of functionTimers) {
                output += `${key}:\n\tCalled: ${value.totalCalls}\n\tTotal Time: ${value.totalDuration}ms` +
                    `\n\tAverage Time: ${value.totalDuration/value.totalCalls}ms\n\n`;
            }

            return output;
        }
    }
}

