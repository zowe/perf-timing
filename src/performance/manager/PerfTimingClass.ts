import { IEnabled, IPerformanceApi } from "../manager/interfaces";
import { PerformanceApi } from "../api/PerformanceApi";
import * as pkgUp from "pkg-up";
import { isArray } from "util";

//////////////////////////////////////////
/////////////// GLOBAL TYPING ////////////
//////////////////////////////////////////
const GLOBAL_SYMBOL = Symbol.for("org.zowe.PerfTimingClass");

// tslint:disable:no-namespace interface-name
// We need to add stuff to the global scope so that one package can manage
// all possible loaded packages. To do this, we need to disable a few rules.

/**
 * Because of how package dependencies work, there could be multiple instances
 * of this package within an application. To avoid each of these packages having
 * a process.exit hook, we add all of them to the global scope under a unique
 * symbol. The first instance created will be the manager of all of these global
 * packages. This allows for a single `process.on('exit')` hook to be registered.
 *
 * @see {@link PerfTimingClass#_savePerformanceResults}
 */
declare namespace NodeJS {
    interface Global {
        [GLOBAL_SYMBOL]: Map<symbol, IPerformanceApi>;
    }
}

declare var global: NodeJS.Global;
// tslint:enable

//////////////////////////////////////////
/////////////// END TYPING ///////////////
//////////////////////////////////////////

/**
 * This class is a manager of all available performance tools
 */
export class PerfTimingClass implements IEnabled {
    // @TODO recommend wrapping stuff since we forward up requests through dummy object
    // to reduce overhead on getApi call
    public static readonly ENV_ENABLED_KEY = "ENABLED";
    public static readonly ENV_PREFIX = "PERF_TIMING";
    public readonly isEnabled: boolean;

    private _instanceSymbol: symbol; // used to uniquely identify the instance

    private _managedApi: PerformanceApi;

    // Document
    constructor() {
        // First check if the environment prefx is set to the ENV_ENABLED_KEY value
        if (
            process.env[PerfTimingClass.ENV_PREFIX] &&
            process.env[PerfTimingClass.ENV_PREFIX].toUpperCase() === PerfTimingClass.ENV_ENABLED_KEY
        ) {
            // The environment was set so performance metrics are enabled.
            this.isEnabled = true;

            // Check if the global scope has been created. If it hasn't been created
            // this is the first manager called so it will become the main manager.
            if (!global[GLOBAL_SYMBOL]) {
                // Create the global map on first run
                global[GLOBAL_SYMBOL] = new Map();
                process.on("exit", () => this._savePerformanceResults());
            }
        } else {
            // Performance was not enabled.
            this.isEnabled = false;
        }
    }

    public getApi(): PerformanceApi {
        if (this._managedApi == null) {
            // Defers the import until it is needed, will improve performance when
            // performance api hasn't yet been called.
            const performanceApi: typeof PerformanceApi = require("../api/PerformanceApi").PerformanceApi;
            this._managedApi = new performanceApi(this);


            // Create a unique entry in the global map
            if (this.isEnabled) {
                const pkg = require(pkgUp.sync());
                this._instanceSymbol = Symbol(`${pkg.name}@${pkg.version}`);

                // Save the managedApi in the global space. It has been typed so that
                // changes to the wrapper methods of IPerformanceTools cannot be easily
                // changed.
                global[GLOBAL_SYMBOL].set(this._instanceSymbol, this._managedApi);
            }
        }

        // Can guarentee that this will be unique per package instance :)
        return this._managedApi;
    }

    private _savePerformanceResults(): void {
        const metrics = global[GLOBAL_SYMBOL].entries();
        const outputMetrics: any = {}; // @TODO proper typing

        // Get timing first to not skew the results
        outputMetrics.nodeTiming = this.getApi().getNodeTiming();

        for(const [key, value] of metrics) {
            const symbolValue = key.toString();

            // Place into an array to handle the case where the same
            // package might have existed twice.

            if (!isArray(outputMetrics[symbolValue])) {
                outputMetrics[symbolValue] = [];
            }

            outputMetrics[symbolValue].push(value.getMetrics());
        }

        // @TODO save to file
        console.dir(outputMetrics, {
            depth: 4,
            colors: true
        });
        console.log(JSON.stringify(outputMetrics)); // @TODO ensure there are no running timers before we print metrics
    }
}
