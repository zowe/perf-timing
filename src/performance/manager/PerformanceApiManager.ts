import { IPerformanceApi, IPerformanceApiManager } from "./interfaces";
import { PerformanceApi } from "../api/PerformanceApi";
import * as pkgUp from "pkg-up";
import { isArray } from "util";
import { ENV_PREFIX, GLOBAL_SYMBOL } from "../../constants";
import { Environment } from "../../environment";

//////////////////////////////////////////
/////////////// GLOBAL TYPING ////////////
//////////////////////////////////////////
// tslint:disable:no-namespace interface-name
// We need to add stuff to the global scope so that one package can manage
// all possible loaded packages. To do this, we need to disable a few rules.

/**
 * Because of how package dependencies work, there could be multiple instances
 * of this package within an application. To avoid each of these packages having
 * a process.exit hook, we add all of them to the global scope under a unique
 * symbol. The first instance created will be the manager of all of these global
 * packages. This allows for a single `process.on('exit')` hook to be registered.
 *
 * @see {@link PerformanceApiManager#_savePerformanceResults}
 */
declare namespace NodeJS {
    interface Global {
        [GLOBAL_SYMBOL]: Map<symbol, IPerformanceApi>;
    }
}

declare var global: NodeJS.Global;
// tslint:enable

/////////////////////////////// ///////////
/////////////// END TYPING ///////////////
//////////////////////////////////////////

/**
 * This class is a manager of all available performance tools
 */
export class PerformanceApiManager implements IPerformanceApiManager {
    // @TODO recommend wrapping stuff since we forward up requests through dummy object
    // @TODO to reduce overhead on getApi call
    public static readonly ENV_ENABLED = `${ENV_PREFIX}_ENABLED`;
    public static readonly ENV_ENABLED_KEY = "TRUE";
    public readonly isEnabled: boolean;
    public readonly packageUUID: string;

    private _instanceSymbol: symbol; // used to uniquely identify the instance

    private _managedApi: PerformanceApi;

    // @TODO Document
    constructor() {
        // First check if the environment prefix is set to the ENV_ENABLED_KEY value
        if (Environment.getValue(PerformanceApiManager.ENV_ENABLED).toUpperCase() === PerformanceApiManager.ENV_ENABLED_KEY) {
            // The environment was set so performance metrics are enabled.
            this.isEnabled = true;

            // Generate the packageUUID
            const pkg = require(pkgUp.sync());
            this.packageUUID = `${pkg.name}@${pkg.version}`;

            // Check if the global scope has been created. If it hasn't been created
            // this is the first manager called so it will become the main manager.
            if (!global[GLOBAL_SYMBOL]) {
                // Create the global map on first run
                global[GLOBAL_SYMBOL] = new Map();
                process.on("exit", () => this._savePerformanceResults());
            }
        } else {
            // Performance was not enabled.
            this.isEnabled = false;
        }
    }

    public getApi(): PerformanceApi {
        if (this._managedApi == null) {
            // Defers the import until it is needed, will improve performance when
            // performance api hasn't yet been called.
            const performanceApi: typeof PerformanceApi = require("../api/PerformanceApi").PerformanceApi;
            this._managedApi = new performanceApi(this);


            // Create a unique entry in the global map
            if (this.isEnabled) {
                this._instanceSymbol = Symbol(this.packageUUID);

                // Save the managedApi in the global space. It has been typed so that
                // changes to the wrapper methods of IPerformanceTools cannot be easily
                // changed.
                global[GLOBAL_SYMBOL].set(this._instanceSymbol, this._managedApi);
            }
        }

        // Can guarantee that this will be unique per package instance :)
        return this._managedApi;
    }

    private async _savePerformanceResults(): Promise<void> {
        const metrics = global[GLOBAL_SYMBOL].entries();
        const outputMetrics: any = {}; // @TODO proper typing

        // Get timing first to not skew the results
        outputMetrics.nodeTiming = this.getApi().getNodeTiming();

        for (const [key, value] of metrics) {
            const symbolValue = key.toString();

            // Place into an array to handle the case where the same
            // package might have existed twice.

            if (!isArray(outputMetrics[symbolValue])) {
                outputMetrics[symbolValue] = [];
            }

            outputMetrics[symbolValue].push(value.getMetrics());
        }

        console.dir(outputMetrics, {
            depth: 4,
            colors: true
        });

        // Require the IO utility at this point to reduce total number of
        // requires in the calling library when performance monitoring
        // is not enabled.
        (await import("../../io/saveMetrics")).saveMetrics(outputMetrics);

        // @TODO ensure there are no running timers before we print metrics
    }
}

// Register the PERF_TIMING_ENABLED variable
Environment.register(PerformanceApiManager.ENV_ENABLED, "");
