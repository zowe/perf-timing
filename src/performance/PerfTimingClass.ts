import { PerformanceTools } from './PerformanceTools';
import { IPerfEnabled } from './interfaces';

//////////////////////////////////////////
/////////////// GLOBAL TYPING ////////////
//////////////////////////////////////////
const GLOBAL_SYMBOL = Symbol.for("org.zowe.PerfTimingClass");

declare module NodeJS {
    interface Global {
        [GLOBAL_SYMBOL]: Map<Symbol, PerformanceTools>;
    }
}

declare var global: NodeJS.Global
//////////////////////////////////////////
/////////////// END TYPING ///////////////
//////////////////////////////////////////

/**
 * This class is a manager of all available performance tools
 */
export class PerfTimingClass implements IPerfEnabled {
    // @TODO recommend wrapping stuff since we forward up requests through dummy object
    // to reduce overhead on getApi call
    public static readonly ENV_PREFIX = "PERF_TIMING";
    private static readonly INSTANCE_SYMBOL = Symbol("PerfTimingClass"); // used to uniquely identify the instance

    public readonly isPerfEnabled: boolean;
    private readonly _isMainManager: boolean;

    private _managedApi: PerformanceTools;

    // Document
    constructor() {
        if (
            process.env[PerfTimingClass.ENV_PREFIX] &&
            process.env[PerfTimingClass.ENV_PREFIX].toUpperCase() === "ENABLE"
        ) {
            this.isPerfEnabled = true;

            if (global[GLOBAL_SYMBOL]) {
                this._isMainManager = false;
            } else {
                this._isMainManager = true;

                // Create the global map on first run
                global[GLOBAL_SYMBOL] = new Map();

                process.on("exit", () => {
                    const metrics = global[GLOBAL_SYMBOL].entries();

                    for(const [key, value] of metrics) {
                        console.log(key);
                        console.log(value.getMetrics());
                    }
                })
            }
        } else {
            this.isPerfEnabled = false;
        }
    }

    public getApi(): PerformanceTools {
        if (this._managedApi == null) {
            // Defers the import until it is needed, will improve performance when 
            // performance api hasn't yet been called.
            const PerfImport: typeof PerformanceTools = require("./PerformanceTools").PerformanceTools;
            this._managedApi = new PerfImport(this);

            if (this.isPerfEnabled) {
                global[GLOBAL_SYMBOL].set(PerfTimingClass.INSTANCE_SYMBOL, this._managedApi);
            }
        }

        // Can guarentee that this will be unique per package instance :)
        return this._managedApi;
    }
}
