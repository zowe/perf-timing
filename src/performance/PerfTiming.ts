import { PerfTimingClass } from "./manager/PerfTimingClass";

/**
 * An instance of a {@link PerfTimingClass}. This object is the only thing
 * intended to be directly imported by any dependents of this package.
 *
 * @example <caption>Usage</caption>
 * ```typescript
 * import { PerfTiming } from "@zowe/perf-timing";
 *
 * if (PerfTiming.isEnabled) {
 *   // Do stuff with PerfTiming
 *   PerfTiming.getApi().mark("start");
 *   PerfTiming.getApi().mark("end");
 *   PerfTiming.getApi().measure("beginning to end", "start", "end");
 * }
 * ```
 */
export const PerfTiming = new PerfTimingClass(); // tslint:disable-line:variable-name
