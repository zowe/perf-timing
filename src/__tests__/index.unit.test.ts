/*!
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

import { PerformanceApiManager } from "../performance/manager";
import { PerfTiming } from "../index";

describe("public exports", () => {
    it("should export a constructed api manager", () => {
        // Simply checks that we've exported the API manager object.
        expect(PerfTiming).toBeInstanceOf(PerformanceApiManager);
    });

    it("should have unchanged public exports", () => {
        expect(require("../index")).toMatchSnapshot();
    });
});
