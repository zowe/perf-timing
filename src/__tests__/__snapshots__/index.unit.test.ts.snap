// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`public exports should have unchanged public exports 1`] = `
Object {
  "PerfTiming": PerformanceApiManager {
    "isEnabled": false,
  },
}
`;
